File: app/android/app/src/main/java/com/jun/weather/viewmodel/WeatherPointViewModel.java
package com.jun.weather.viewmodel;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.jun.weather.repository.AppRepository;
import com.jun.weather.repository.db.FileRepository;
import com.jun.weather.util.CLogger;
import com.jun.weather.viewmodel.entity.WeatherPointModel;

import java.util.List;

public class WeatherPointViewModel extends CustomViewModel<WeatherPointModel> {
    public WeatherPointViewModel(@NonNull Application application, AppRepository repository) {
        super(application, repository);

        LiveData<List<FileRepository.ShortWeatherPoint>> shortWeatherPointList = repository.getShortWeatherPointList();
        shortWeatherPointList.observeForever((value) -> {
            WeatherPointModel weatherPointModel = new WeatherPointModel();
            String befKey = "";
            int startIdx = -1, endIdx = -1;
            for(int i = 0; i < value.size(); i++) {
                FileRepository.ShortWeatherPoint point = value.get(i);
                WeatherPointModel.WeatherPoint model = new WeatherPointModel.WeatherPoint();
                model.deg_1 = point.getDeg1();
                model.deg_2 = point.getDeg2();
                model.deg_3 = point.getDeg3();
                model.midTempCode = point.getMidTempCode();
                model.midWeatherCode = point.getMidWeatherCode();
                model.x = point.getX();
                model.y = point.getY();
                weatherPointModel.weatherPointList.add(model);

                if(befKey.equals(point.getDeg1()+" "+point.getDeg2())) {
                    endIdx++;
                } else {
                    if(!befKey.isEmpty()) {
                        weatherPointModel.indexMap.put(befKey, new Integer[]{startIdx, endIdx});
                    }
                    startIdx = i;
                    endIdx = i;
                    befKey = point.getDeg1()+" "+point.getDeg2();
                }
            }

            if(startIdx != -1 && startIdx == endIdx) {
                weatherPointModel.indexMap.put(befKey, new Integer[]{startIdx, value.size()-1});
            }
            appExecutor.runInUIIO(() -> data.setValue(weatherPointModel));
        });
    }

    public void updatePointData(Context context) {
        appExecutor.runInRepositoryIO(() -> repository.loadShortWeatherPointList(context));
    }
}

File: app/android/app/src/main/java/com/jun/weather/viewmodel/WeatherPointViewModel.kt
package com.jun.weather.viewmodel

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.jun.weather.repository.AppRepository
import com.jun.weather.ui.entity.WeatherPoint
import com.jun.weather.ui.entity.WeatherPointModel
import com.jun.weather.util.PreferenceUtils
import kotlinx.coroutines.launch

class WeatherPointViewModel(repository: AppRepository) : BaseViewModel(repository) {
    private val _data = MutableLiveData<WeatherPointModel>()
    val data: LiveData<WeatherPointModel> = _data

    private val _selectedPoint = MutableLiveData<WeatherPoint>()
    val selectedPoint = _selectedPoint

    val pointArr = Array(3) {
        MutableLiveData<WeatherPoint>()
    }

    init {
        updatePointArr()
    }

    private fun updatePointArr() {
        PreferenceUtils.getInstance().pointList.forEachIndexed { index, weatherPoint ->
            pointArr[index].postValue(weatherPoint)
        }
    }

    fun updatePointData(context: Context) {
        viewModelScope.launch {
            val value = repository.getShortWeatherPointList(context)

            val weatherPointModel = WeatherPointModel()
            var befKey = ""
            var startIdx = -1
            var endIdx = -1
            value.forEachIndexed { index, model ->
                weatherPointModel.weatherPointList.add(
                        WeatherPoint(
                                model.deg1,
                                model.deg2,
                                model.deg3,
                                model.x,
                                model.y,
                                model.midTempCode,
                                model.midWeatherCode,
                        )
                )
                if (befKey == "${model.deg1} ${model.deg2}") {
                    endIdx++
                } else {
                    if (!befKey.isEmpty()) {
                        weatherPointModel.indexMap[befKey] = arrayOf(startIdx, endIdx)
                    }
                    startIdx = index
                    endIdx = index
                    befKey = "${model.deg1} ${model.deg2}"
                }
            }
            if (startIdx != -1 && startIdx == endIdx) {
                weatherPointModel.indexMap[befKey] = arrayOf(startIdx, value.size - 1)
            }
            _data.postValue(weatherPointModel)
        }
    }

    fun postSelectedPoint(point : WeatherPoint) {
        PreferenceUtils.getInstance().addPoint(point)
        _selectedPoint.postValue(point)
        updatePointArr()
    }
}